### PR TITLE
[MLIR][OpenMP] Add `OpenMP_Clause` tablegen definitions

### PR DESCRIPTION
**Pull request by skatrak**
This patch adds a new tablegen file for the OpenMP dialect containing the list of clauses currently supported.

### CODE REVIEW

File: mlir/include/mlir/Dialect/OpenMP/OpenMPClauses.td [hunk 1]
```diff
@@ -0,0 +1,1183 @@
...
+  > : OpenMP_Clause</*isRequired=*/false, traits, arguments, assemblyFormat,
+                    description, extraClassDeclaration> {
+  let traits = [
+    ReductionClauseInterface
+  ];
+
+  let arguments = (ins
+    Variadic<OpenMP_PointerLikeType>:$reduction_vars,
+    OptionalAttr<SymbolRefArrayAttr>:$reductions,
+    UnitAttr:$byref
```
[the remark this answers is not in the document]
Comment (skatrak) [reply mlir/include/mlir/Dialect/OpenMP/OpenMPClauses.td]: Thanks for pointing it out. I've rebased the PR stack and addressed this. I ended up adding the byref information to all reduction-like clauses (reduction, taskreduction and inreduction) and adding it to the custom parser/printer for ReductionVarList in other patches of the stack.

**Review by tblah — APPROVED** [review]
LGTM, thanks for the refactoring.

File: mlir/include/mlir/Dialect/OpenMP/OpenMPClauses.td [hunk 2]
```diff
@@ -0,0 +1,1184 @@
...
+#define OPENMP_CLAUSES
+
+include "mlir/Dialect/OpenMP/OpenMPOpBase.td"
+
+//===----------------------------------------------------------------------===//
+// V5.2: [5.11] `aligned` clause
+//===----------------------------------------------------------------------===//
+
+class OpenMP_AlignedClauseSkip<
+    bit traits = false, bit arguments = false, bit assemblyFormat = false,
```
Comment (bhandarkar-pranav) [inline mlir/include/mlir/Dialect/OpenMP/OpenMPClauses.td:29]: Any reason you have dropped the `skip` prefix on the template arguments?

File: mlir/include/mlir/Dialect/OpenMP/OpenMPClauses.td [hunk 3]
```diff
@@ -0,0 +1,1184 @@
...
+  }];
+
+  let description = [{
+    The `alignment_values` attribute additionally specifies alignment of each
+    corresponding aligned operand. Note that `aligned_vars` and
+    `alignment_values` should contain the same number of elements.
+  }];
+}
+
+def OpenMP_AlignedClause : OpenMP_AlignedClauseSkip<>;
```
Comment (bhandarkar-pranav) [inline mlir/include/mlir/Dialect/OpenMP/OpenMPClauses.td:50]: The PR generally looks good to me, but i have a style nit which can be termed a preference. This line here is not sitting right with me (unless I am missing something), `OpenMP_AlignedClause` is a `OpenMP_AlignedClauseSkip` with its default template arguments. But `OpenMP_AlignedClauseSkip` does not skip the inheritance of anything in its default state. Or am I missing something obvious?

**Review by bhandarkar-pranav — APPROVED** [review]
LGTM - just a couple style questions

**Comment by skatrak** [comment]
Thank you @bhandarkar-pranav for taking a look! Since both of your comments are related, I'll try to explain the reasoning here. Hopefully it makes sense to you.

Basically the idea is that we want to have `def`s for each clause that do not skip anything, which will be the main ones used while defining operations, but we also want the ability to skip parts of them in a very small amount of edge cases. That's what the separation is for, which I think we are both onboard with.

Then there is the definition of the `OpenMP_XyzClauseSkip` classes, which is the one that seems to be causing the confusion. How these are supposed to be read is "An Xyz clause skipping the following fields: ...". So, when we do `def OpenMP_XyzClause : OpenMP_XyzClauseSkip<>` we're saying: "Define an Xyz clause skipping none of the fields". I think this is more readable especially when actually skipping fields while defining operations. For example:
```
def TaskloopOp : OpenMP_Op<"taskloop", traits = [ ... ],
    clauses = [ OpenMP_ReductionClauseSkip<extraClassDeclaration = true, ...> ]> { ... }
```
In that case, we're saying that the `omp.taskloop` operation takes a reduction clause but we want to skip the `extraClassDeclaration` field (and possibly others) inherited from it.

I think keeping the "skip" prefix on the template parameters would make things redundant (e.g. `OpenMP_XyzClauseSkip<skipDescription = true>`). Alternatively, we could make the base class use the regular clause name: `OpenMP_XyzClause<skipDescription = true>`, which was my first idea. The disadvantage of this is that then we'd have two options that in my opinion would be worse because they make the much more common case more verbose:
- Renaming the definition that includes every field into something like `OpenMP_XyzClause{Base,Default,NoSkip}`.
- Instantiating the template directly in the clause list every time: `traits = [ OpenMP_Xyz1Clause<>, OpenMP_Xyz2Clause<skipDescription = true> ]`

In the end, I thought that `traits = [ OpenMP_Xyz1Clause, OpenMP_Xyz2ClauseSkip<description = true> ]` was the most readable alternative of the options I considered.

**Comment by bhandarkar-pranav** [comment]
> Thank you @bhandarkar-pranav for taking a look! Since both of your comments are related, I'll try to explain the reasoning here. Hopefully it makes sense to you.
> 
> Basically the idea is that we want to have `def`s for each clause that do not skip anything, which will be the main ones used while defining operations, but we also want the ability to skip parts of them in a very small amount of edge cases. That's what the separation is for, which I think we are both onboard with.
> 
> Then there is the definition of the `OpenMP_XyzClauseSkip` classes, which is the one that seems to be causing the confusion. How these are supposed to be read is "An Xyz clause skipping the following fields: ...". So, when we do `def OpenMP_XyzClause : OpenMP_XyzClauseSkip<>` we're saying: "Define an Xyz clause skipping none of the fields". I think this is more readable especially when actually skipping fields while defining operations. For example:
> 
> ```
> def TaskloopOp : OpenMP_Op<"taskloop", traits = [ ... ],
>     clauses = [ OpenMP_ReductionClauseSkip<extraClassDeclaration = true, ...> ]> { ... }
> ```
> 
> In that case, we're saying that the `omp.taskloop` operation takes a reduction clause but we want to skip the `extraClassDeclaration` field (and possibly others) inherited from it.
> 
> I think keeping the "skip" prefix on the template parameters would make things redundant (e.g. `OpenMP_XyzClauseSkip<skipDescription = true>`). Alternatively, we could make the base class use the regular clause name: `OpenMP_XyzClause<skipDescription = true>`, which was my first idea. The disadvantage of this is that then we'd have two options that in my opinion would be worse because they make the much more common case more verbose:
> 
> * Renaming the definition that includes every field into something like `OpenMP_XyzClause{Base,Default,NoSkip}`.
> * Instantiating the template directly in the clause list every time: `traits = [ OpenMP_Xyz1Clause<>, OpenMP_Xyz2Clause<skipDescription = true> ]`
> 
> In the end, I thought that `traits = [ OpenMP_Xyz1Clause, OpenMP_Xyz2ClauseSkip<description = true> ]` was the most readable alternative of the options I considered.

Thank you, this makes sense.

**Review by mjklemm — APPROVED** [review]
LGTM.

**Pull request merged**